### ISSUE TITLE
Fixing a merge conflict

### ISSUE THREAD

**Issue by arginite** [archimatetool/archi-modelrepository-plugin#57]
It's unclear how a merge conflict can currently be resolved without loss of data.

Two users edit a model that results in a merge conflict.
The last user to the party is notified that a merge conflict exists and is greeted by a diff of XML generated by Archi.
The Archi generated XML isn't impossible to follow but resolving anything but the simplest of conflicts will be very very time consuming as Archi generated XML isn't very human readable.

Currently/Going forward is there a strategy/plan for dealing with merges?

**Comment by Phillipus** [comment]
We are now working on this.

**Comment by Phillipus** [comment]
Ideas for making this easier in the UI are welcome now. We need to gather requirements and suggestions for this.

- XML merging as it is now could remain as an option for "expert" mode.
- Are we aiming for something like a "structural compare"? This is similar to how diffs can be represented in Eclipse for code changes:

![compare](https://user-images.githubusercontent.com/600504/45028692-f08bfd80-b03d-11e8-9f82-1ff18640f960.png)

**Comment by Phillipus** [comment]
At the moment, the options are choosing between "ours" or "theirs"  changes. Anything more fine grained such as selective changes will be more complicated.

**Comment by arginite** [comment]
The structural compare would be a great first step, being able to go deeper into each structure that has changed allowing for more fine grain merges would be a good final goal.

**Comment by tjsiefring** [comment]
What about the ability to create branches from within the plugin instead of everyone working from master?  Then you could follow the more typical pull request and merge into master. As in the example above, if the two users are working from different branches they wouldn't have conflicts.  At least not until the branches merge.  At that point it might make sense to manage the merge outside of Archi.

**Comment by tjsiefring** [comment]
Looks like this and #24 are the same or rather this would be solved the adding the new branching feature.

**Comment by Phillipus** [comment]
We'll take a look at branches as well

**Comment by jbsarrodie** [comment]
Hi, I finally found some time to describe what I think would be a good target.

Of course, the first step when conflict happen would be to explain the user that the publication is not possible because of some conflicts:
![archi collaboration - conflicts_1](https://user-images.githubusercontent.com/5757396/45715428-8bcfb780-bb94-11e8-8a93-2261881ccca9.jpg)

In this very basic dialog user can either abort or decide to solve conflicts, but a third way could be to offer the choice to automatically solve conflicts by always keeping user changes or remote changes. I personnally don't like this kind of "simple" choices because if people don't pay attention they will certainly make mistakes at some point.

That's why I suggest to design a simple but effective dialog to solve conflicts:
![archi collaboration - conflicts_2](https://user-images.githubusercontent.com/5757396/45715621-f2ed6c00-bb94-11e8-92f5-a7ce58294633.jpg)

This UI is based on common UI we can find for code conflicts resolution (main difference being that we don't show common ancestor as I don't think this makes much sens in our context):
* We have on the top the list of concepts or views that are in conflict. Here we don't simply show the filename of XML files, but the actual ArchiMate type and concept/view name in the model. This will help people to easily understand what is in conflict. In addition we add a column that indicates if a choice has been done (the choice being between "mine" version and "theirs" version.
* When selecting one of the conflicts, the botom part of the dialog is updated accordingly to show some details about the choosen concept:
  - Type
  - Name
  - Documentation
  - Properties
  - Type and name of source concept (for a relationship)
  - Type and name of target concept (for a relationship)
* These details are shown for the version of the current user's model on the left ("mine") and for the HEAD of remote ("theirs").
* User can then compare both version and choose either "mine" or "theirs". When doing so, the top list of conflicts is updated to reflect the choice.
* When all conflicts has been solved, then "Proceed" button becomes active (was unactive before) and user can continue.

If the selected conflict is not a concept but a view, then bottom panels should show (in addition to type, name, documentation and properties) the view itself:
![archi collaboration - conflicts_3](https://user-images.githubusercontent.com/5757396/45716227-968b4c00-bb96-11e8-91ea-20776a84fd7c.jpg)


For this to work, this means that we have to:
1. Identify concepts and views in conflict
2. For each of thos concept/view, get type and name to display the list of conflicts inUI
2. Open the remote's HEAD's version of the model in Archi (but don't show it in UI). This involves the same mecanism than "Extract model from this commit)
3. When the user select a conflict, get type, name, documentation, and properties) of associated concept/view in both models (the one used by the user and the one extracted from remote's HEAD) and show those informations in bottom panes.
4. If selected concept is a view, then generate an image from both version of the view and show those images in bottom panes (or maybe when the user clic on some "show view image" button somewhere).

I'd suggest to:
1. start by focusing on top level list (that should list real typ and name instead of XML filename)
2. implement bottom panes without view image (only view informations)
3. implement view image.

Does this make sens to you ?

**Comment by jbsarrodie** [comment]
>You mentioned "Open the remote's HEAD's version of the model in Archi (but don't show it in UI). This involves the same mecanism than "Extract model from this commit)". Are you suggesting to create another model ("theirs") for the comparison? I.e not to use the conflicted "ours" and "theirs" XML files?

Exactly, which leads to the sequence you described in your latest comment:
>1. We load "our" model and "their" model into memory by extracting and loading the contents of "our" latest commit and "their" remote commit.
>2. We access the JGit MergeResult to provides us with a list of XML files (EObjects) that are conflicted.
>3. We can get the ID of each conflicted XML file EObject. From this ID we can find the real EObject in "our" or "their" model.

**Comment by jbsarrodie** [comment]
FWIW, I've just tested the new branch: it looks really good and promising!

**Comment by Phillipus** [comment]
> FWIW, I've just tested the new branch: it looks really good and promising!

Needs more work - refactoring, use models to get Eobjects, better layout.

I'm thinking of better ways to display the Views. Not much room there!

**Comment by jbsarrodie** [comment]
>use models to get Eobjects

Not sure to understand

>better layout
>I'm thinking of better ways to display the Views. Not much room there!

I'm working on it too...

**Comment by Phillipus** [comment]
> Not sure to understand

At the moment, I don't fully do stage (3) above. Will do that next.

At this stage I just want to get the ours and theirs info and display it, even if it does look ugly.

**Comment by jbsarrodie** [comment]
Here's a proposition for a (hopefully) improved UI:
![mockup conflicts](https://user-images.githubusercontent.com/5757396/46170873-4ea7ab80-c29f-11e8-949e-d5efd8f69a81.png)

"Main" and "Properties" tabs could then be similar to the one we have in Archi (thus intuitive for the user). The "View" tab (only visible for views of course) would simply contain the image.

Any remarks?

**Comment by Phillipus** [comment]
Funny you should say that! I've been trying out different UI arrangements all today and tabs are what I'm now playing with. ;-)

My experiments show:

- If the selected object is not a View you can fit all information in one panel (name, doc, properties, source, target)
- If the selected object is a View you really need as much space as possible to display the image, so it needs a new tab or area to show it
- Tabs do solve some problems but the user is constantly switching between them (not all information can be seen at once)

I wonder typically how many objects will be in conflict? How much tap dancing the user will do?

**Comment by jbsarrodie** [comment]
>If the selected object is not a View you can fit all information in one panel (name, doc, properties, source, target)

Not sure one panel is the best. Documentation can be rather long and should the last thing to show in a tab as you don't know how long it is, but same could be said for properties. Maybe having something that mimics the original "Main" and "Properties" tab in Archi would be more intuitive (and manages long docs or prop list).

>If the selected object is a View you really need as much space as possible to display the image, so it needs a new tab or area to show it

Yes, that's why I put tabs on top and not on the left like in Archi itself.

>I wonder typically how many objects will be in conflict? How much tap dancing the user will do?

In my usage (small team of 2-3 people) I faced approx. 5 conflicts in a full year. I don't expect users to face conflicts very often if they publish often and if work is well spread across the team.

>Tabs do solve some problems but the user is constantly switching between them (not all information can be seen at once)

What is done now can be seen as the first version to stabilize (which is way better than our first MVp from last year).

A (maybe) better option would be to compare each attributes and props between "mine " and "theirs" and only show differences. For a view, if no attribute (name, documentation) nor props differ, then this means that the differences are in the diagram, thus we show it. With this logic, user would see only differences, not everything.

A slightly different approach would be to find differences (as just described) but emphazise them. So user could see everything, but impacted tabs could have a star to mark that something is in conflict, and changed attributes or props could be written on a (light) red background.

**Comment by Phillipus** [comment]
Yes, I thought about about only showing differences. But perhaps the user might want to be reminded about all of the information for an object. For example, if there is no documentation displayed because of no difference they might say "what was this object? I need to see the documentation to remember".

**Comment by Phillipus** [comment]
I will do a tabbed version for testing. BTW we only need one set of tabs, not two as in your mockup. Thus selecting one tab shows the diffs for both objects.

**Comment by Phillipus** [comment]
> That's an option I though about yesterday too. Not sure about the UX, but that can be tested for sure.

If there are two sets of tabs, one for "theirs" and one for "ours" the user will have to click twice to compare like-for-like. However, If the splitter is inside one set of tabs then clicking on a tab will show both related sides.

Can you think of any advantages for having two sets of tabs?

**Comment by jbsarrodie** [comment]
>Can you think of any advantages for having two sets of tabs?

Well... Er... No ;-)

**Comment by Phillipus** [comment]
Tabbed interface committed.

**Comment by jbsarrodie** [comment]
For those who follow this issue, here's a screenshot of this new version:
![screenshot from 2018-09-29 20 23 24](https://user-images.githubusercontent.com/5757396/46249236-0a92e300-c426-11e8-8fba-83270c11268e.png)

That's big improvement !

I only have two remarks:

It might be interesting to be able to zoom in/out in the view tab. Maybe to keep in mind for a future version (but not needed in this sprint)

From a user experience point of view, I think that the position of the tabs is not optimal: being on the left under the "Mine" button, it is not obvious that tabs are for both versions. I'd move buttons for "Mine" and "Theirs" below the whole content:
![screenshot from 2018-09-29 20 23 24-2](https://user-images.githubusercontent.com/5757396/46249355-20090c80-c428-11e8-8fa1-9bb19655032f.png)

What do you think?

**Comment by jbsarrodie** [comment]
Another remark... we have to think about two exceptions: access and influence. Both have additional attributes that can be in conflict (read/write type for access, and strength for influence) and should be displayed.

**Comment by Phillipus** [comment]
> It might be interesting to be able to zoom in/out in the view tab.

Yes, I agree. I've added some scales sliders now, one for each side. I figured that you might want to zoom out on either image.

> I'd move buttons for "Mine" and "Theirs" below the whole content

Done. BTW - you can also set "mine" and "theirs" in the top table (combo box in cell).

> we have to think about two exceptions: access and influence.

Done.

**Comment by jbsarrodie** [comment]
Really great !

It seems we have a final version for this sprint ;-)

**Comment by Phillipus** [comment]
Let's test fully. ;-)

**Comment by Phillipus** [comment]
I'm testing this with a larger model (The Aviation one from TOG).

A "Refresh Model" action has a bottle-neck of about 15 seconds when the MergeConflictHandler extracts the two models (ours and theirs from the commits). The "extractModel" method run once take about 7-8 seconds. So x2 = about 15 seconds.

**Comment by jbsarrodie** [comment]
Why reloading 'ours' which is already loaded in UI?

**Comment by Phillipus** [comment]
> Why reloading 'ours' which is already loaded in UI?

True. This assumes that our model == master. Will there ever be a case where this is not so?

**Comment by Phillipus** [comment]
> Why reloading 'ours' which is already loaded in UI?

RevertCommitAction also uses the MergeConflictHandler and that doesn't use the model in the UI.

**Comment by jbsarrodie** [comment]
>Will there ever be a case where this is not so?

Well, by definition, 'mine/ours' is the model on which we work, so it will be always be master (or any other branch when we'll have implemented branching).

From a UX point of view, I would suggest to have a simple dialog that appears when conflicts exist, so that user can decide to solve them or not. Then, only if he/she decides to solve them, we load 'theirs' with a progress bar and then open the new conflict solver.

**Comment by jbsarrodie** [comment]
>RevertCommitAction also uses the MergeConflictHandler and that doesn't use the model in the UI.

But when is "RevertCommitAction" used ? If I remember well it used to be used to revert a commit other than the last one, but we no longer allow this (as it was almost always raising conflicts).

**Comment by Phillipus** [comment]
It's used as part of "RevertCommitsAction" which is "Revert all up to this commit". The RevertCommand is called and then any Merge conflicts have to be resolved.

**Comment by jbsarrodie** [comment]
>It's used as part of "RevertCommitsAction" which is "Revert all up to this commit". The RevertCommand is called and then any Merge conflicts have to be resolved.

So I have a simple fix : remove "Revert all up to this commit" ;-)

In practice, I have never seen it used; users always use "Restore to this commit" which doesn't "pollute" the history (and never generate conflicts)

**Comment by Phillipus** [comment]
> So I have a simple fix : remove "Revert all up to this commit" ;-)

Well, it would make our life easier. ;-)

But let's look at the core problem here for merge conflicts when pulling:

1. The MergeResult provides us with a list of grafico XML files that conflict (ours and theirs)
2. Each XML file provides enough information about the eObject by itself, but does not resolve referenced eObjects. Certainly when we want to create an image of a View we need the whole model to resolve everything.
3. So we extract the grafico model(s) from their commit and our commit, and load the full model via GraficoModelImporter

The time-consuming part is extracting all the grafico XML files from a given commit:

            try(TreeWalk treeWalk = new TreeWalk(repository)) {
                treeWalk.addTree(commit.getTree());
                treeWalk.setRecursive(true);

                while(treeWalk.next()) {
                    ObjectId objectId = treeWalk.getObjectId(0);
                    ObjectLoader loader = repository.open(objectId);
                    
                    File file = new File(tmpFolder, treeWalk.getPathString());
                    file.getParentFile().mkdirs();
                    
                    try(FileOutputStream out = new FileOutputStream(file)) {
                        loader.copyTo(out);
                    }
                }
            }

**Comment by Phillipus** [comment]
I've committed more changes:

1. "Our" model is simply the current model already loaded. This will probably affect "Revert all up to this commit" action so this action will need to be removed.
2. Extracting "Their" model still takes about 7 seconds on my PC with the Aviation model. I've put a progress monitor on it so the user can cancel the operation and revert back to previous local state. I can't see how to optimise this part as many XML files are being written (see above).

> From a UX point of view, I would suggest to have a simple dialog that appears when conflicts exist, so that user can decide to solve them or not. Then, only if he/she decides to solve them, we load 'theirs' with a progress bar and then open the new conflict solver.

That's possible but wouldn't the user be curious as to what the conflicts are? If so then this dialog would become unnecessary?

**Comment by tjsiefring** [comment]
Hi guys, where did this end up? Is a release planned?  I'm interested in this and the branching features.

**Comment by Phillipus** [comment]
Yes, we made progress on this so that the conflict can be seen visually and logically (not as a xml text diff). We're aiming to release this when the branching feature has matured.

**Comment by jassoks** [comment]
@Phillipus can you please suggest how do I get this capability (UI to resolve conflicts during branch merge) in my Archi 4 installation ?

**Comment by Phillipus** [comment]
> @Phillipus can you please suggest how do I get this capability (UI to resolve conflicts during branch merge) in my Archi 4 installation ?

It is implemented in the latest version of the coArchi collaboration plug-in.
